src/main/java/org/hmhb/program/DefaultProgramService.java

```diff
@@ -69,16 +74,169 @@ public Program getById(
     ) {
         LOGGER.debug("getByIds called: ids={}", ids);
         requireNonNull(ids, "ids cannot be null");
-        return dao.findByIdIn(ids);
+        return dao.findAll(ids);
     }
 
     @Timed
     @Override
     public List<Program> getAll() {
         LOGGER.debug("getAll called");
-        List<Program> target = new ArrayList<>();
-        dao.findAll().forEach(target::add);
-        return target;
+        return dao.findAllByOrderByNameAsc();
+    }
+
+    @Timed
+    @Override
+    public List<Program> search(
+            @Nonnull MapQuery query
+    ) {
+        LOGGER.debug("search called: query={}", query);
+        requireNonNull(query, "query cannot be null");
+
+        MapQuerySearch search = query.getSearch();
+
+        if (search == null) {
+            return getAll();
+        }
+
+        if (search.getProgram() != null) {
+
+            return Collections.singletonList(
+                    getById(search.getProgram().getId())
+            );
+
+        }
+
+        County county = search.getCounty();
+        Organization organization = search.getOrganization();
+        ProgramArea programArea = search.getProgramArea();
+
+        Long countyId = null;
+
+        if (county != null) {
+            countyId = county.getId();
+        }
+
+        Long organizationId = null;
+
+        if (organization != null) {
+            organizationId = organization.getId();
+        }
+
+        Long programAreaId = null;
+
+        if (programArea != null) {
+            programAreaId = programArea.getId();
+        }
+
+        /*
+         * Both of these potential implementations (nested ifs and flattened ifs) are really ugly,
+         * but using JPA Specification looks much more complicated.
+         *
+         * For now, with only 3 search params, I'm just going to have all permutations of
+         * searches in the DAO.  If anymore params are added and the permutations get worse,
+         * I'll have to use JPA Specification or figure something else out.
+         */
+//        return doSearchFlattenedIfs(organizationId, countyId, programAreaId);
```

Comment:
I went ahead and merged this in last night with both implementations in here, and using the nested one.
Now they are in here part of history, so if we decide, they can be recovered.

Based on the code here and the linked articles, I think creating a `Specification` for each and using them in here would probably be the cleanest.  I agree that we probably don't want to drag in Querydsl.

Something like the below pseudo code would scale if more criteria are added, and reads cleaner IMO:

```
        Specifications specifications = where(ALWAYS_TRUE);

        if (organizationId != null) {
            specifications = specifications.and(
                    hasOrganizationId(organizationId)
            );
        }

        if (countyId != null) {
            specifications = specifications.and(
                    hasCountyId(countyId)
            );
        }

        if (programAreaId != null) {
            specifications = specifications.and(
                    hasProgramAreaId(programAreaId)
            );
        }

        return dao.findAll(specifications);
```